docs/developer-docs/smart-contracts/advanced-features/http-gateways.mdx

```diff
@@ -0,0 +1,105 @@
+
+### Terminology
+
+- **Request**: A message sent to a server or endpoint.
+
+- **Client**: A service that is able to send an HTTP formatted request and receive a response from a server.
+
+- **Gateway**: Enables the transfer of inbound and outbound messages.
+
+## HTTP gateway workflow
+
+On ICP, an HTTP request is handled by an HTTP client using the following workflow:
+
+1. An HTTP client makes an outbound request.
+
+2. The HTTP gateway intercepts the request and resolves the canister ID of the request's destination.
+
+3. The request is encoded with Candid and sent in a query call to the destination canister's `http_request` method.
+
+4. ICP's API gateway intercepts this request, resolves the canister ID based on the hostname address, and then forwards the request to the relevant replica node. All requests made to ICP through an HTTP gateway are forwarded through this API gateway.
+
+5. The canister receives and processes the request, then returns the response.
+
+6. The HTTP gateway decodes the Candid encoding on the response and sends the request again via an update call to the canister's `http_request_update` method. The update call goes through consensus on the subnet.
+
+7. If necessary, the HTTP gateway fetches additional response body data through query calls.
+
+8. The HTTP gateway validates the response's certificate, if applicable.
+
+9. The HTTP gateway returns the decoded response to the HTTP client.
+
+Additional details can be found in the [HTTP gateway specification](https://internetcomputer.org/docs/current/references/http-gateway-protocol-spec).
+
+## Running a local HTTP gateway
+
+To run your own local HTTP gateway, a proof-of-concept implementation can be used that enables a secure end-to-end connection with canisters deployed on ICP.
+
+This implementation features:
+
+- Translation between HTTP asset requests and ICP API calls.
+
+- Detects ICP domains from principals and custom domain records.
+
+- Terminates TLS connections locally.
+
+- Bypasses remote gateway denylists.
+
+- Resolves crypto domains.
+
+### Prerequisites
+
+- Download and install [yarn package manger](https://yarnpkg.com/).
+
+- Download and install [Node.js](https://nodejs.org/en).
+
+
+- #### Step 1: Clone the GitHub repo:
+
+```
+git clone https://github.com/dfinity/http-proxy.git
+```
+
+- #### Step 2: Install the project's dependencies:
+
+```
+cd http-proxy
+npm install --global yarn
+yarn set version 3.5.0
+yarn install
+```
+
+- #### Step 3: Build and run the project:
+
+```
+yarn build
+yarn start
+```
+
+You should see output that indicates the local HTTP gateway is ready to use:
+
+```
+{"level":30, "time": "2024-06-13T22:41:28.962Z","pid":41696,"hostname":"JessieMgeonsMBP.attlocal.net","name":"IC HTTP Proxy UI","msg":"Preparing interface"}
+{"level":30, "time": "2024-06-13T22:41:28.981Z","pid":41696,"hostname":"JessieMgeonsMBP.attlocal.net","name":"IC HTTP Proxy UI","msg":"Interface is ready"}
+```
```

Comment:
Let's show how to use this gateway now, and let's enable the user to see that it works. E.g., what port is it running on? Something like a `curl` command to demonstrate that something is happening (like fetching an HTTP point of the frontend dapp, or something similar - ideally some URL that stays stable, maybe just '/')

Reply:
Could you provide some commands on how to achieve that?

Reply:
Followed up on slack